File: macos/Sources/NoctusCore/Services/NavigationHistory.swift
import Foundation

public final class NavigationHistory {
    private var history: [PathRef]
    private var position: Int

    public init(_ initialLocation: PathRef) {
        self.history = [initialLocation]
        self.position = 0
    }

    public var current: PathRef { history[position] }
    public var canGoBack: Bool { position > 0 }
    public var canGoForward: Bool { position < history.count - 1 }

    public func push(_ location: PathRef) {
        guard location != current else { return }
        if position < history.count - 1 {
            history.removeSubrange((position + 1)...)
        }
        history.append(location)
        position = history.count - 1
    }

    public func goBack() -> PathRef {
        precondition(canGoBack, "No back history")
        position -= 1
        return current
    }

    public func goForward() -> PathRef {
        precondition(canGoForward, "No forward history")
        position += 1
        return current
    }
}

File: src/Explorer.Core/Services/NavigationHistory.cs
using NoctusExplorer.Core.Models;

namespace NoctusExplorer.Core.Services;

public sealed class NavigationHistory
{
    private readonly List<PathRef> _history = [];
    private int _position;

    public NavigationHistory(PathRef initialLocation)
    {
        _history.Add(initialLocation);
        _position = 0;
    }

    public PathRef Current => _history[_position];

    public bool CanGoBack => _position > 0;

    public bool CanGoForward => _position < _history.Count - 1;

    public void Push(PathRef location)
    {
        if (location == Current) return;

        // Truncate forward history
        if (_position < _history.Count - 1)
            _history.RemoveRange(_position + 1, _history.Count - _position - 1);

        _history.Add(location);
        _position = _history.Count - 1;
    }

    public PathRef GoBack()
    {
        if (!CanGoBack)
            throw new InvalidOperationException("No back history available.");
        _position--;
        return Current;
    }

    public PathRef GoForward()
    {
        if (!CanGoForward)
            throw new InvalidOperationException("No forward history available.");
        _position++;
        return Current;
    }
}
